### PR TITLE
feat(FEC-10709, FEC-10712): player visibility - Auto-pause when player is out of view, Autoplay only when player is in view

### PR DESCRIPTION
**Pull request by RoyBregman**
### Description of the Changes

In continuity of the floating mechanism (Floating Player - Design & Tasks breakdown ) some additional functionality is now 
required to enhance behaviour when player visibility state changes (Due to scrolling / tabbing in/out)

- added "inview" to autoplay options
- added new autopause config to playback
- changed MultiMap to support all types as keys

solves FEC-10709, FEC-10712

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/event/event-type.js [hunk 1]
```diff
@@ -247,7 +247,19 @@ const CustomEventType: PKEventTypes = {
...
+   */
+  TAB_VISIBILITY_CHANGE: 'tabvisibilitychange',
+  /**
+   * Fired when the player gets in /out of scroll view
+   */
+  SCROLL_VISIBILITY_CHANGE: 'scrollvisibilitychange',
+  /**
+   * Fired when the player changes its visibility state - either by tabbing in / out or scrolling in /out of view
+   */
+  VISIBILITY_CHANGE: 'visibilitychange'
```
Comment (OrenMe) [inline src/event/event-type.js:262]: I still don't see why we need to add seperate events, all say the same at the end of the day.
Either check with Kava if tab is same as scroll - i believe it does, or add the change to include a reason with tab/scroll - which I think is not really needed.

File: src/auto-play-type.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// @flow
+const AutoPlayType: PKAutoPlayTypes = {
+  true: true,
+  IN_VIEW: 'inview'
+};
```
Comment (OrenMe) [inline src/auto-play-type.js:5]: false?

File: src/player.js [hunk 3]
```diff
@@ -399,6 +407,47 @@ export default class Player extends FakeEventTarget {
...
+   * @private
+   */
+  _isVisible: boolean;
+
+  /**
+   * Whether the player is in floating mode
+   * @type {boolean}
+   * @private
+   */
+  _isFloating: boolean = false;
```
Comment (OrenMe) [inline src/player.js:443]: if core player is aware of floating state which is from a plugin we are breaking isolation of features

File: src/player.js [hunk 4]
```diff
@@ -170,12 +171,19 @@ export default class Player extends FakeEventTarget {
...
+   */
+  _mediaEventManager: EventManager;
   /**
    * The event manager of the player.
    * @type {EventManager}
    * @private
    */
-  _eventManager: EventManager;
+  _playerEventManager: EventManager;
+
```
Comment (OrenMe) [inline src/player.js:186]: why two seperate event managers?

**Review by OrenMe — CHANGES_REQUESTED** [review]
Copnsider seperating the entire visibility managment to seperate manager in its own module.

File: src/event/event-type.js (hunk 1 above)
Comment (RoyBregman) [reply src/event/event-type.js:262]: For example the floating feature is only interested of scrolling in / out and not tabbing in/out.
I guess other features might be interested of just tabbing.
I don't see a reason why not giving all options open

File: src/auto-play-type.js (hunk 2 above)
Comment (RoyBregman) [reply src/auto-play-type.js:5]: I added

File: src/player.js (hunk 3 above)
Comment (RoyBregman) [reply src/player.js:443]: Okay... But this is part of the product request...
How else can I implement this?
Why is this breaking isolation? I am not relying of the plugin to be part. Either it is and then I consider it or not.

File: src/event/event-type.js (hunk 1 above)
Comment (OrenMe) [reply src/event/event-type.js:262]: So let's use reason

File: src/player.js (hunk 3 above)
Comment (RoyBregman) [reply src/player.js:443]: it is like saying not to check PIP because it is implemented by the browser.

File: src/player.js (hunk 4 above)
Comment (RoyBregman) [reply src/player.js:186]: The current event manager we have is cleaned and attached every media switch.
We need an event manager that is kept through all the player instance.

File: src/player.js (hunk 3 above)
Comment (OrenMe) [reply src/player.js:443]: Core doesn't know about floating so now it needs to.
If you want to have some generic way to signal core player to ignore condition then this is a generic api that can then be used for other plugins if they want to tell the feature to not trigger

File: src/player.js (hunk 4 above)
Comment (OrenMe) [reply src/player.js:186]: So see comment about separating to its own manager, we do same in other managers.

File: src/player.js (hunk 3 above)
Comment (OrenMe) [reply src/player.js:443]: And per pip, it is not the same as we implement pip internally while floating is not

File: src/player.js [hunk 5]
```diff
@@ -1897,6 +2059,19 @@ export default class Player extends FakeEventTarget {
...
   }
 
+  /**
+     }
+   * Checks auto play configuration and handles initialization accordingly.
+   * @returns {void}
+   * @private
+   */
+  _initAutoPlay(): void {
+    this._posterManager.show();
```
Comment (yairans) [inline src/player.js:2069]: why you changed the previous logic which shows the poster only for autoplay false?
maybe it could cause a flicker

File: src/player.js [hunk 6]
```diff
@@ -1725,6 +1863,16 @@ export default class Player extends FakeEventTarget {
     }
   }
 
+  /**
+   * Update the floating active state.
+   * @param {boolean} isFloating - is player floating.
+   * @returns {void}
+   * @public
+   */
+  changeFloatingState(isFloating: boolean): void {
```
Comment (yairans) [inline src/player.js:1872]: `toggleFloatingState` ? or just `floatingState` setter?

File: src/player.js [hunk 7]
```diff
@@ -1837,33 +1985,47 @@ export default class Player extends FakeEventTarget {
...
-            }
-          } else {
-            onAutoPlayFailed();
+  _initAutoPause(): void {
+    if (this._config.playback.autopause === true) {
+      this._mediaEventManager.listen(this, CustomEventType.VISIBILITY_CHANGE, (e: FakeEvent) => {
+        if (!e.payload.visible) {
+          if (!this._isFloating && !this.isInPictureInPicture() && this._playbackStart && !this.paused) {
+            this.pause();
+            this._autoPaused = true;
```
Comment (yairans) [inline src/player.js:1998]: I would reset it in playing for when the play and pause called by API
(paused on invisible => played by api => paused by api => become visible)

File: src/player.js [hunk 8]
```diff
@@ -399,6 +407,47 @@ export default class Player extends FakeEventTarget {
...
+   * @private
+   */
+  _isTabVisible: boolean;
+
+  /**
+   * Whether the player browser tab is active and in the scroll view
+   * @type {boolean}
+   * @private
+   */
+  _isVisible: boolean;
```
Comment (yairans) [inline src/player.js:436]: is not `_isVisibleInScroll && _isTabVisible`?

File: src/playkit.js [hunk 9]
```diff
@@ -33,6 +33,8 @@ import {AdBreakType} from './ads/ad-break-type';
 import {AdTagType} from './ads/ad-tag-type';
 import {AdEventType} from './ads/ad-event-type';
 import {ScreenOrientationType} from './screen-orientation-type';
+import 'intersection-observer';
```
Comment (yairans) [inline src/playkit.js:36]: why it's needed?

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/player.js (hunk 5 above)
Comment (Yuvalke) [reply src/player.js:2044]: and add also isAudio to this condition, the same as we have  in the master today

File: src/utils/viewability-manager.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+   * @param {Function} listener - the callback to be invoked when visibility is changed (and when starting to observe). The callback is called with a boolean param representing the visibility state
+   * @returns {void}
+   */
+  observe(target: HTMLElement, threshold: number, listener: Function): void {
+    const options = {
+      threshold: threshold
+    };
+    const intersectionObserver = new window.IntersectionObserver((entries: Array<window.IntersectionObserverEntry>) => {
+      const visible = entries[0].intersectionRatio >= threshold;
+      (listener: Function).call(this, visible);
```
Comment (Yuvalke) [inline src/utils/viewability-manager.js:25]: why don't just call the listener?

File: src/player.js [hunk 11]
```diff
@@ -1528,6 +1574,83 @@ export default class Player extends FakeEventTarget {
...
+  get isTabVisible(): boolean {
+    return this._isTabVisible;
+  }
+
+  /**
+   * Gets the player visibility in scroll view state
+   * @returns {boolean} - whether the player is visible in the scroll view
+   * @public
+   */
+  get isVisibleInScroll(): boolean {
```
Comment (Yuvalke) [inline src/player.js:1642]: why do you need the getter and _isVisibleInScroll if you don't have the event SCROLL_VISIBILITY_CHANGE anymore?

Comment (Yuvalke) [reply src/player.js:1642]: why do we need those getters? user can use events for this state changes

**Review by Yuvalke — CHANGES_REQUESTED** [review]

File: src/player-config.js [hunk 12]
```diff
@@ -58,7 +58,8 @@ const DefaultConfig = {
         engine: 'flash',
         format: 'hls'
       }
-    ]
+    ],
+    visibilityThreshold: 50
```
[the remark this answers is not in the document]
Comment (Yuvalke) [reply src/player-config.js:62]: I'm not sure if playback is the right place for this config

**Review by Yuvalke — CHANGES_REQUESTED** [review]

File: src/player-config.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/player-config.js:62]: We talked about it at the DR, you are right, but do you have a better suggestion?

File: src/player.js (hunk 11 above)
Comment (RoyBregman) [reply src/player.js:1642]: the getter isVisibleInScroll is really not relevant anymore. I removed it.
The other getters are needed to get the initial state

Comment (Yuvalke) [reply src/player.js]: the initial state won't be always yes?

File: src/event/event-type.js (hunk 1 above)
Comment (RoyBregman) [reply src/event/event-type.js]: Not relevant anymore. Code moved to Kaltura. Only new event will be dispatched from Kaltura - VISIBILITY_CHANGE

File: src/player.js (hunk 3 above)
Comment (RoyBregman) [reply src/player.js]: Not relevant anymore. Code was refactored

File: src/player.js (hunk 4 above)
Comment (RoyBregman) [reply src/player.js]: Not relevant anymore. Reverted back to one manager

File: src/auto-play-type.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+// @flow
+const AutoPlayType: PKAutoPlayTypes = {
+  true: true,
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/auto-play-type.js]: Fixed and change TRUE to ALWAYS

File: src/event/event-manager.js [hunk 14]
```diff
@@ -12,7 +12,7 @@ import FakeEvent from './fake-event';
  * @implements {IDestroyable}
  */
 class EventManager {
-  _bindingMap: MultiMap<Binding_> | null;
+  _bindingMap: MultiMap<string, Binding_> | null;
```
Comment (Yuvalke) [inline src/event/event-manager.js:15]: why did you change it?

**Review by Yuvalke — CHANGES_REQUESTED** [review]

**Comment by Yuvalke** [comment]
Please change the PR comment and run docs:generate cause you added a new API

File: src/player.js [hunk 15]
```diff
@@ -171,11 +172,12 @@ export default class Player extends FakeEventTarget {
   }
 
   /**
-   * The event manager of the player.
+   * The event manager of the current media.
```
Comment (OrenMe) [inline src/player.js:175]:
```suggestion
   * The event manager of the player.
```

File: src/player.js [hunk 16]
```diff
@@ -461,7 +463,7 @@ export default class Player extends FakeEventTarget {
         this._posterManager.setSrc(this._config.sources.poster);
         this._handleDimensions();
         this._handlePreload();
-        this._handleAutoPlay();
+        this._initAutoPlay();
```
Comment (OrenMe) [inline src/player.js:466]: if we don't change anything let's keep current name

File: src/auto-play-type.js [hunk 17]
```diff
@@ -0,0 +1,8 @@
+// @flow
+const AutoPlayType: PKAutoPlayTypes = {
+  ALWAYS: true,
```
Comment (OrenMe) [inline src/auto-play-type.js:3]: keep it true, as we don't call false NEVER

File: src/event/event-type.js [hunk 18]
```diff
@@ -251,7 +251,11 @@ const CustomEventType: PKEventTypes = {
   /**
    * Fired when the drm license is responded from the DRM server
    */
-  DRM_LICENSE_LOADED: 'drmlicenseloaded'
+  DRM_LICENSE_LOADED: 'drmlicenseloaded',
+  /**
+   * Fired when the player changes its visibility state - either by tabbing in / out or scrolling in /out of view
+   */
+  VISIBILITY_CHANGE: 'visibilitychange'
```
Comment (OrenMe) [inline src/event/event-type.js:258]: why do we need this here? we can pull it to kaltura-player and merge it in the events getter

File: src/player-config.js [hunk 19]
```diff
@@ -78,7 +78,11 @@ const DefaultConfig = {
   drm: {
     keySystem: ''
   },
-  network: {}
+  network: {},
+  visibility: {
+    tolerance: 10,
+    playerThreshold: 50
+  }
```
Comment (OrenMe) [inline src/player-config.js:85]: move to kaltura-player

File: src/player-config.js [hunk 20]
```diff
@@ -78,7 +78,11 @@ const DefaultConfig = {
   drm: {
     keySystem: ''
   },
-  network: {}
+  network: {},
+  visibility: {
+    tolerance: 10,
```
Comment (OrenMe) [inline src/player-config.js:83]: i think this name is not to clear, and it will be more correct to get this as an array cause sometimes people will want non linear stops

File: src/player.js [hunk 21]
```diff
@@ -1902,35 +1908,25 @@ export default class Player extends FakeEventTarget {
...
-        } else {
-          if (capabilities.mutedAutoPlay) {
-            if (this.muted && !this._fallbackToMutedAutoPlay) {
-              onMutedAutoPlay();
-            } else if (allowMutedAutoPlay) {
-              onFallbackToMutedAutoPlay();
-            } else {
-              onAutoPlayFailed();
-            }
+  autoPlay(): void {
```
Comment (OrenMe) [inline src/player.js:1911]: it is not correct to call this autoPlay, as it is more correct to say this is a non user gesture play AKA programmatic play.
WDYT to put this in the play API and pass to it an object {programmatic: boolean} and if it is yes then call `_handleAutoPlay` and return.
BTW, the check for audio should still show the poster in this case.

File: src/auto-play-type.js (hunk 17 above)
Comment (RoyBregman) [reply src/auto-play-type.js]: We could call it OFF but ok, reverted to TRUE :)

File: src/player.js (hunk 16 above)
Comment (OrenMe) [reply src/player.js:466]: @RoyBregman ? is this change necessary?

File: src/player.js (hunk 21 above)
Comment (OrenMe) [reply src/player.js:1911]: @RoyBregman ?

File: src/player.js (hunk 16 above)
Comment (RoyBregman) [reply src/player.js:466]: No, it is not, reverted back to original name

File: src/player-config.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/player-config.js]: fixed

File: src/player.js (hunk 5 above)
Comment (RoyBregman) [reply src/player.js]: fixed

File: src/player.js (hunk 6 above)
Comment (RoyBregman) [reply src/player.js]: not relevant anymore

File: src/player.js (hunk 8 above)
Comment (RoyBregman) [reply src/player.js]: not relevant anymore

File: src/playkit.js (hunk 9 above)
Comment (RoyBregman) [reply src/playkit.js]: ie11 support. moved to viewabilityManager

File: src/utils/viewability-manager.js (hunk 10 above)
Comment (RoyBregman) [reply src/utils/viewability-manager.js]: fixed

File: src/player.js (hunk 11 above)
Comment (RoyBregman) [reply src/player.js]: Not relevant anymore

File: src/event/event-type.js (hunk 18 above)
Comment (RoyBregman) [reply src/event/event-type.js]: done

File: src/player-config.js (hunk 19 above)
Comment (RoyBregman) [reply src/player-config.js]: done

File: src/player-config.js (hunk 20 above)
Comment (RoyBregman) [reply src/player-config.js]: Not relevant anymore

File: src/player.js (hunk 21 above)
Comment (RoyBregman) [reply src/player.js]: done

File: src/player.js [hunk 22]
```diff
@@ -517,10 +517,16 @@ export default class Player extends FakeEventTarget {
 
   /**
    * Start/resume playback.
+   * @param {boolean} programmatic - indicates the play call was not user gestured and should be handled like auto play.
    * @returns {void}
    * @public
    */
-  play(): void {
+  play(programmatic: boolean = false): void {
```
Comment (OrenMe) [inline src/player.js:524]: I think adding a true/false boolean to API is not clear, this is why I suggested to add it as an object so when one calls the API it is clear what is the action: `player({programmatic: true})`

File: src/player.js (hunk 21 above)
Comment (OrenMe) [reply src/player.js]: please see https://github.com/kaltura/playkit-js/pull/521#discussion_r551856918 - this comment is still relevant and not handled.

> BTW, the check for audio should still show the poster in this case.

was this checked? I guess it should be OK as on initial load if it is audio we will get here and show it - please make sure this is covered and make sure it is a tested use case in QA plan

Comment (RoyBregman) [reply src/player.js]: I'm not sure I understand the comment - audio case logic is remained the same. So is autoplay true and false.
The only change in logic here is that I show poster also for autoplay "inview".
This is done because someone can define threshold as 90% and scroll slowly until he reaches that point. We don't want to see a black player in the mean while.

File: src/player.js [hunk 23]
```diff
@@ -517,12 +517,12 @@ export default class Player extends FakeEventTarget {
 
   /**
    * Start/resume playback.
-   * @param {boolean} programmatic - indicates the play call was not user gestured and should be handled like auto play.
+   * @param {PKPlayOptionsObject} playOptions - additional options to control the play. i.e. programmatic true indicates the play call was not initiated by a user gesture and should be handled like auto play.
```
Comment (OrenMe) [inline src/player.js:520]: see example on hot to describe object properties with jsdoc https://stackoverflow.com/questions/6460604/how-to-describe-object-arguments-in-jsdoc

Comment (RoyBregman) [reply src/player.js]: done

**Review by OrenMe — APPROVED** [review]

File: flow-typed/types/auto-play-types.js [hunk 24]
```diff
@@ -0,0 +1,2 @@
+// @flow
+declare type PKAutoPlayTypes = { [type: string | boolean]: string | boolean };
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply flow-typed/types/auto-play-types.js]: yes, fixed

File: src/player.js [hunk 25]
```diff
@@ -1837,33 +1985,47 @@ export default class Player extends FakeEventTarget {
...
-              onMutedAutoPlay();
-            } else if (allowMutedAutoPlay) {
-              onFallbackToMutedAutoPlay();
-            } else {
-              onAutoPlayFailed();
-            }
-          } else {
-            onAutoPlayFailed();
+  _initAutoPause(): void {
+    if (this._config.playback.autopause === true) {
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/player.js]: done

File: src/player.js (hunk 7 above)
Comment (RoyBregman) [reply src/player.js]: Not relevant anymore

**Review by OrenMe — APPROVED** [review]

**Review by OrenMe — DISMISSED** [review]

**Review by OrenMe — APPROVED** [review]

**Review by Yuvalke — APPROVED** [review]
